README.md
# udp_rx
This is still currently pre-1.0 software, breaking changes are likely to be made until a 1.0 release

udp_rx is a program created to tunnel UDP traffic through a TLS 1.2+ connection. It is useful for securing old protocols that cannot be updated to tcp or for some reason, can't use dtls. If you use this tool you **should** firewall off the UDP ports used by the protocol to prevent redirection by a malicious third party and **should** firewall off the UDP port used by udp_rx.

Please see gen_keys_readme.txt for creating/using SSL keys with udp_rx

Please see IPTABLES_RULE.txt for a sample IPTABLES command to firewall off a port protected by udp_rx

## Export Control Notice
Otis udp_rx software has been designed to utilize information security technology described in the Category 5 – Part 2 of the Commerce Control List, within Part 774 of the Export Administration Regulations (“EAR”)(15 CFR 774).  However, the Otis udp_rx software has been made publicly available in accordance with Part 742.15(b) of the EAR and is therefore not subject to U.S. export regulations. 

Before downloading this software, be aware that the country in which you are located may have restrictions related to the import, download, possession, use and/or reexport of encryption items.  It is your responsibility to comply with any applicable laws and regulations pertaining the import, download, possession, use and/or reexport of encryption items.

## How to use it
You have a UDP packet that you want to be sent to `192.168.1.250` with a destination port of `4444`. The data field of that packet is 

`[5,4,3,2,1]`

Take that packet and prepend a udp_rx header to it according to the specification in `header_format.md`. In this case the header would look like:

`[0x75,0x00,0x01,0x13,0xc4,0x7c,0x04,0xC0,0xa8,0x01,0xFA,0x80]`

and the resulting packet would be:

`[0x75,0x00,0x01,0x13,0x11,0x5c,0x04,0xC0,0xa8,0x01,0xFA,0x80,5,4,3,2,1]`

Send that packet to `localhost:55555` (or whatever port udp_rx is configured to listen to)

udp_rx will attempt to create a TLS connection to another instance of udp_rx running on `192.168.1.250`. If it succeeds:

udp_rx will receive the packet and perform two different actions depending on platform.

### Linux
udp_rx will receive the packet and craft a UDP packet from `sender:src_port` to `destination:dest_port` as if udp_rx was never in the middle, so that it can be received by the desination application.

### Windows
udp_rx will receive the packet and create a UDP packet from `localhost:random_port` to `desination:dest_port`. It will prepend the sender's IP address and src_port to the data. So if the sender was `192.168.1.100`, from port `4445` the packet data field would be:

```[192,168,1,100,11,5D,10,9,8,7,6,5,4,3,2,1]```

## License
This program is released under the MIT License. For details, please see the LICENSE file

## Building
Please see `BUILD.md` for instructions on building udp_rx.

## Credits
This program uses the following open source libraries
* [logrus](https://github.com/sirupsen/logrus) - Copyright Simon Eskildsen (MIT License)
* [lumberjack](https://github.com/natefinch/lumberjack/tree/v2.1) - Copyright Nate Finch (MIT License)

---

Copyright 2018, Otis Elevator Company

CHANGELOG.MD
# udp_rx Changelog
## AAA
This is a 1.0 release. 

* added flag for log file location: `logpath`
* default log path on Linux is now: `/var/log/udp_rx`
* Default log path on Windows is now: `c:\programdata\udp_rx\udp_rx.log`

## A25
**This is still pre-1.0**
Fixed a bug where duplicate IPTables rules could be created by `udprx_firewall` and would not be deleted by `udprx_firewall -unset` 

udp_rx.go
// Copyright 2018 Otis Elevator Company. All rights reserved.
// Use of this source code is govered by the MIT license which
// can be found in the LICENSE file.

// Author: Jeremy Mill: jeremy.mill@otis.com

// Otis udp_rx software has been designed to utilize information
// security technology described in the Category 5 – Part 2 of the
// Commerce Control List, within Part 774 of the Export Administration
// Regulations (“EAR”)(15 CFR 774).  However, the Otis udp_rx software
// has been made publicly available in accordance with Part 742.15(b)
// of the EAR and is therefore not subject to U.S. export regulations.
// Before downloading this software, be aware that the country in which
// you are located may have restrictions related to the import, download,
// possession, use and/or reexport of encryption items.  It is your
// responsibility to comply with any applicable laws and regulations
// pertaining the import, download, possession, use and/or reexport of
// encryption items.

package main

import (
	"crypto/tls"
	"flag"
	"fmt"
	"os"
	"time"

	"./udprxlib"

	log "github.com/sirupsen/logrus"
	lumberjack "gopkg.in/natefinch/lumberjack.v2"
)

// Version is a constant that is this verion of the code, according to OTIS standards
const Version = "AAA31825AAA"

// RemoteTLSPort is the port of the remote TLS server (also the port of the local TLS server)
const RemoteTLSPort = ":55554"

// tls config
var clientConf *tls.Config
var serverConf *tls.Config

// const startup arg values
const defaultListenAddr = ""

var confFilePath = "/etc/udp_rx/udp_rx_conf.json"
var defaultKeyPath = "/etc/udp_rx/udp_rx.key"
var defaultCertPath = "/etc/udp_rx/udp_rx.crt"
var defaultCACertPath = "/etc/udp_rx/ca.crt"
var defaultLogPath = "/var/log/udp_rx.log"

var listenAddr, keyPath, certPath, caCertPath string

func main() {
	fmt.Printf("Starting udp_rx at: %s\n", time.Now())

	// modify the defaults if we're on windows
	if isWindows() {
		modifyDefaultsWindows()
	}
	// get and parse command line args
	versionFlag := flag.Bool("version", false, "Print the Version number and exit")
	logFlag := flag.Int("loglevel", 0, "level of logging. 0 is warn+, 1 is Info+, 2 is debug+")
	listenAddrFlag := flag.String("bindaddr", defaultListenAddr, "The IP address to bind the listening UDP socket to")
	cpuprofileFlag := flag.String("cpuprofile", "", "If specified writed a cpuprofile to the given filename")
	maxProfilingPacketsFlag := flag.Int("maxprofpackets", 1000, "the maximum number of packets allowed to be forwarded during CPU profiling")
	netProfilingFlag := flag.Bool("netprof", false, "turn on net profiling")
	lumberjackFlag := flag.Bool("lumberjack", false, "use lumberjack local file logging")
	logPathFlag := flag.String("logpath", defaultLogPath, "the path to log to")
	// configuration filepath override
	confFileFlag := flag.String("conf", confFilePath, "Override the default configuration filepath")
	// certificate flags
	keyPathFlag := flag.String("keypath", defaultKeyPath, "Override the default key path/name which is ./keys/server.key")
	certPathFlag := flag.String("certpath", defaultCertPath, "Override the default certificate path/name which is ./server.crt")
	caCertPathFlag := flag.String("cacert", defaultCACertPath, "Set the Certificate Authority Certificate to add to the trust")
	flag.Parse()
	// init the logger
	if *lumberjackFlag {
		fmt.Println("using lumberjack")
		log.SetOutput(&lumberjack.Logger{
			Filename:   *logPathFlag,
			MaxSize:    500, // megabytes
			MaxBackups: 3,
			MaxAge:     28,   // days
			Compress:   true, // disabled by default
		})
	}
	fmt.Printf("Logger configured at: %s\n", time.Now())
	// if version flag, print version and exit
	if *versionFlag {
		fmt.Printf("Version is: %s\n", Version)
		os.Exit(0)
	}
	// load config file
	conf, err := udprxlib.ParseConfig(*confFileFlag)
	if err == nil {
		setConfigValues(&conf, listenAddrFlag, keyPathFlag, certPathFlag, caCertPathFlag)
	} else {
		log.Warn("Error parsing the config file. Error: ", err.Error())
		setConfigValues(nil, listenAddrFlag, keyPathFlag, certPathFlag, caCertPathFlag)
	}

	configLogger(logFlag)
	log.Warning("Starting udp_rx version: ", Version)

	// check if CPU profiling is on
	if *cpuprofileFlag != "" {
		log.Warning("Turning on cpu profling...")
		udprxlib.EnableCPUProfiling(*maxProfilingPacketsFlag, cpuprofileFlag)
	}
	// set net profiling
	if *netProfilingFlag {
		if *maxProfilingPacketsFlag != 1000 {
			log.Warning("overriding number of profiling packets to ", *maxProfilingPacketsFlag)
		}
		udprxlib.EnableNetProfiling(*maxProfilingPacketsFlag)
	}

	// This will block until the year is > 1970
	log.Debug("Blocking until time > 1970")
	for {
		ctime := time.Now()
		if ctime.Year() > 1970 {
			break
		}
		time.Sleep(100 * time.Millisecond)
	}
	log.Debug("Time sync done, unblocking")

	// load server cert as tls certs
	log.Debug("keypath: ", certPath)
	log.Debug("certpath: ", keyPath)
	cer, err := tls.LoadX509KeyPair(certPath, keyPath)
	if err != nil {
		log.Fatal(err)
	}
	// load the CA into the trusted store and return it for serverconf
	rootCAs := udprxlib.ConfigureRootCAs(&caCertPath)
	// configure ssl
	clientConf = &tls.Config{
		RootCAs:      rootCAs,
		Certificates: []tls.Certificate{cer},
	}
	serverConf = udprxlib.GetServerConfig(rootCAs, &cer)

	// start listening on the UDP port in go routine
	udpListenerDone := make(chan error, 1)
	go udprxlib.UDPListener(&listenAddr, clientConf, udpListenerDone)
	// start listening on TCP on main thread (blocking main from returning)
	tcpListenerDone := make(chan error, 1)
	udprxlib.TCPListener(&listenAddr, serverConf, tcpListenerDone)
}

func configLogger(logFlag *int) error {
	log.SetFormatter(&log.JSONFormatter{})
	if *logFlag == 0 {
		log.SetLevel(log.WarnLevel)
		log.Warn("LogLevel set to warn")
	} else if *logFlag == 1 {
		log.SetLevel(log.InfoLevel)
		log.Info("LogLevel set to Info")
	} else {
		log.SetLevel(log.DebugLevel)
		log.Debug("LogLevel set to Debug")
		udprxlib.ForwardMap = make(map[string]int)
	}
	return nil
}

func setConfigValues(conf *udprxlib.ConfFile, listAddrArg, keyPathArg, certPathArg, caCertPathArg *string) {
	// precedence is: command line arg -> config file -> program default
	// listen addr
	if *listAddrArg != defaultListenAddr {
		listenAddr = *listAddrArg
	} else if conf != nil && conf.ListenAddr != defaultListenAddr {
		listenAddr = conf.ListenAddr
	} else {
		listenAddr = defaultListenAddr
	}

	// key
	if *keyPathArg != defaultKeyPath {
		keyPath = *keyPathArg
	} else if conf != nil && conf.KeyPath != defaultKeyPath {
		keyPath = conf.KeyPath
	} else {
		keyPath = defaultKeyPath
	}

	// cert
	if *certPathArg != defaultCertPath {
		certPath = *certPathArg
	} else if conf != nil && conf.CertPath != defaultCertPath {
		certPath = conf.CertPath
	} else {
		certPath = defaultCertPath
	}

	// ca cert
	if *caCertPathArg != defaultCACertPath {
		caCertPath = *caCertPathArg
	} else if conf != nil && conf.CaCertPath != defaultCACertPath {
		caCertPath = conf.CaCertPath
	} else {
		caCertPath = defaultCACertPath
	}

}

func modifyDefaultsWindows() {
	confFilePath = "c:\\programdata\\udp_rx\\udp_rx_conf.windows.json"
	defaultKeyPath = "c:\\programdata\\udp_rx\\udp_rx.key"
	defaultCertPath = "c:\\programdata\\udp_rx\\udp_rx.crt"
	defaultCACertPath = "c:\\programdata\\udp_rx\\ca.crt"
	defaultLogPath = "c:\\programdata\\udp_rx\\udp_rx.log"
}

func isWindows() bool {
	return os.PathSeparator == '\\' && os.PathListSeparator == ';'
}
